src/lib/utils/object-graph.ts

```diff
@@ -0,0 +1,172 @@
+import { v4 as uuidv4 } from 'uuid'
+
+/**
+ * Represents a serialized object structure where circular/repeated references
+ * are replaced with `{ __refId }` and the full objects are provided in the
+ * `include` map.
+ */
+export interface FlattenedGraph<T> {
+  data: T
+  include: Record<string, unknown>
+}
+
+/**
+ * A reference to an object stored in the `include` section of a
+ * `DereferencedData` structure.
+ */
+export interface Reference {
+  __refId: string
+}
+
+/**
+ * Serializes an object containing circular and/or repeated references into a
+ * JSON-safe format by extracting duplicates into an `include` map.
+ */
+export function flattenObjectGraph<T>(input: T): FlattenedGraph<T> {
+  // Map of all seen objects. Each object gets a UUID and a flag for whether
+  // it's reused.
+  const seen = new Map<unknown, { __refId: string; isDuplicate: boolean }>()
+
+  // The final store of extracted objects that will be referenced by `{ __refId }`
+  const include: FlattenedGraph<T>['include'] = {}
+
+  // Guards against infinite recursion by tracking active __refIds during
+  // serialization
+  const serializing = new Set<string>()
+
+  /**
+   * First pass: detects duplicates and assigns UUIDs to all objects. Only marks
+   * objects as `isDuplicate` if seen more than once.
+   */
+  function findDuplicates(value: unknown) {
+    if (typeof value !== 'object' || value === null) return
+
+    const existing = seen.get(value)
```

Comment:
Oh interesting, the object itself is the key. I didn't know you could do that with JavaScript Maps!

Reply:
I see it in the MDN docs now. I'm always serializing objects into string keys by hand 😂😭